lib/registry/registry.rb

```diff
@@ -138,17 +126,33 @@ def blob(repo, digest, outpath = nil)
       end
     end
 
-    def digest(repo, tag)
-      tag_path = "/v2/#{repo}/manifests/#{tag}"
-      dohead(tag_path).headers[:docker_content_digest]
-    rescue DockerRegistry2::InvalidMethod
-      # Pre-2.3.0 registries didn't support manifest HEAD requests
-      doget(tag_path).headers[:docker_content_digest]
+    def digest(repo, tag, architecture = nil, os = nil)
+      manifest = manifest(repo, tag)
+      json_body = JSON.parse(manifest.body)
+
+      # Multi-arch images
+      if json_body.key?('manifests')
+        manifests = json_body['manifests']
+
+        raise DockerRegistry2::InvalidInput, 'Both architecture and os must be defined because this is a Multi-arch image' if architecture.nil? || os.nil?
+
```

Comment:
@deitch 

I need some help here with the design of parsing  `Multi-arch images` 

There a couple of decisions we need to make here

1. If there are multiple entries what should happen? 

I am currently filtering based on the new `architecture` and `os` parameters. if so: 

- What should happen if the user does not pass the os and  architecture parameters? I believe that we should throw an error as configured
- What should happen if there are no matches? I am currently returning `nil` but I think that should also throw an error there


- Return a custom object with with architecture, os, and digest?

- Return the raw manifests?

Reply:
> no matches

yeah, I would think that is an error. Maybe a custom one, so that someone can compare a generic error (e.g. could not reach the registry or auth) vs specifically could not match the requested platform.

> What should happen if the user does not pass the os and architecture parameters? I believe that we should throw an error as configured

I would think we should return the digest of the top-level object (likely an index).

Reply:
I refactored it with the latest commit. Changes include: 

1. Throws `NotFound`  when it reaches the end of the manifest and it did not find anything
2. if architecture and os are not specified, it returns the first entry
3. if architecture is defined and os isn't, it returns the first entry that matches the architecture
4. if os isn't defined and architecture is, it returns the first entry that matches the os
5. Otherwise, it checks that both os and architecture  matches and returns that

Reply:
> Throws NotFound when it reaches the end of the manifest and it did not find anything

👍 

> if architecture and os are not specified, it returns the first entry

And why not the index itself? The ordering is arbitrary, so it will be inconsistent between repos, or even tags of the same repo.

> if architecture is defined and os isn't, it returns the first entry that matches the architecture
> if os isn't defined and architecture is, it returns the first entry that matches the os

I think it would be better to have both or none. The actual OCI and docker stuff use "platform", which is a combination `os/arch[/variant]`, e.g. `linux/amd64` or `linux/arm64/v8`. Omitting just OS or arch returns an error (variant os optional).

Reply:
> And why not the index itself? The ordering is arbitrary, so it will be inconsistent between repos, or even tags of the same repo.

Is this appropriate in Ruby? I understand that we did not define any type, but I'd be surprised if a method that normally returns a string returns a object all of sudden

> I think it would be better to have both or none. The actual OCI and docker stuff use "platform", which is a combination os/arch[/variant], e.g. linux/amd64 or linux/arm64/v8. Omitting just OS or arch returns an error (variant os optional).

That's fair. Should we support variant as well?

Reply:
I updated and simplified it

it returns the `manifests` index if : 

- architecture and os are not specified
- architecture is defined and os isn't
- os isn't defined and architecture is

Otherwise, it checks that both os and architecture matches and returns the digest if found. Otherwise it throws `DockerRegistry2::NotFound`

Reply:
Added supports for `variant` as well